### ISSUE TITLE
DL4J/SameDiff: Add easy way of performing (softmax) temperature scaling calibration

### ISSUE THREAD

**Issue by AlexDBlack** [deeplearning4j/deeplearning4j#6961]
Related: https://github.com/deeplearning4j/deeplearning4j/issues/4565

http://alondaks.com/2017/12/31/the-importance-of-calibrating-your-deep-model/

The softmax aspect isn't too difficult, but there's also the optimization of the temperature parameter T. That's also not too hard in theory, but it would be nice to have built in tools to optimize T.

**Comment by tschut** [comment]
To get the same result you can scale values during sampling. For my own project I implemented this: https://github.com/keras-team/keras/blob/b2771d13b0f9146cde43af00f5a5156f56614436/examples/lstm_text_generation.py#L67. 

This scales the output of your softmax layer according to a temperature `t`. After that you can select from `dist[]` in the usual way.

Real quick and dirty:
```
    private void scale(double[] dist, double t) {
        double[] exp_dist = new double[dist.length];
        double exp_sum = 0.0;
        for (int i = 0; i < dist.length; i++) {
            dist[i] = Math.log(dist[i]) / t;
            exp_dist[i] = Math.exp(dist[i]);
            exp_sum += exp_dist[i];
        }
        for (int i = 0; i < dist.length; i++) {
            dist[i] = exp_dist[i] / exp_sum;
        }
    }
```

**Comment by agibsonccc** [comment]
Closing — SameDiff has been completely rewritten with a new DSP (Dynamic Shape Plan) execution engine, native plan compilation, CUDA graph replay, and Triton compilation. The architecture referenced in this issue no longer exists. If you still experience this issue with the upcoming release, please open a new issue.
